### PR TITLE
Realize with externs

### PR DESCRIPTION
**Pull request by zvookin**
This pull request allows passing a list of C function pointers or Func's, each paired with a name, to satisfy extern dependencies of a JIT compilation request. This allows running realize or infer_bounds without relying on dlsym finding an extern by name in the calling process and also customizing the list of functions that are passed in.

### CODE REVIEW

File: src/JITModule.cpp [hunk 1]
```diff
@@ -512,31 +512,93 @@ void JITModule::make_externs(const std::vector<JITModule> &deps, llvm::Module *m
 }
 
 void *JITModule::main_function() const {
-    if (!defined()) {
```
Comment (dsharletg) [inline src/JITModule.cpp]: How about replacing the if defined check with an assert/internal_assert?

Comment (zvookin) [inline src/JITModule.cpp]: It is no longer possible to construct a JITModule without initializing JITModuleContents and there is no way to reset the IntrusivePtr so it seems a worthless assert.

**Comment by dsharletg** [comment]
Would it make sense to change (or add) an interface to set/add externs to a Func for jitting, instead of passing them everywhere they're needed? I.e. instead of:

    f.realize(..., { { "extern", fn } });

It might be:

    f.add_extern_definition("extern", fn);
    // or
    f.add_extern_definitions({ { "extern", fn } });

There's a few reasons for thinking this:

- To make the interface more consistent with the other JIT relevant interfaces (`set_error_handler`, etc.). It's a little surprising that some extern function pointers are supplied by setting them ahead of a call to `realize`, while others are passed as a list to realize. (BTW: I smell a nice refactoring once this goes in to remove the old JIT handlers stuff and just use this extern functionality instead).
- It feels a little off to have to pass the externs map to multiple calls. Suppose you are using `compile_jit`, followed by `realize`: do both calls need the externs mapping? What about multiple calls to `realize`? 
- Multiple optional arguments make a function slightly annoying to use.

Anyways, I generally avoid stateful interfaces where possible, but this case seems one where it makes sense.

**Comment by zvookin** [comment]
It is definitely part of my goal to get rid of the handlers setting functionality on Func. It has always struck me as confusing, it requires too many lines of code in too many places for each handler, and results in issues with packaging code that can be used both for JIT and AOT. 

After the pending Pipeline changes, the state would be carried on Pipeline, not Func. I'm not sure this materially changes anything, but Pipeline may be enough of an API evolution to consider killing all the handlers and moving to just setting externs on Pipeline. This might be a slight loss of usability for some simple cases perhaps, but if one has handlers packaged into a library of functionality one uses with a lot of Halide code, it is probably easier as all handlers can be passed in via one structure. For calibration, I think we'd document this by adding text to HalideRuntime explaining how to provide a list of externs to JIT.

The high level idea with this branch was to have a thing where all the dependencies of a compilation could be provided to the actual compilation. Hence the design choice to pass the list of externs to all routines which JIT. This seems like a different thing than having the code itself set it's dependency. They're both useful, though I'd probably steer people toward bundling up all the deps into a compilation step as opposed to trying to provide them when building the Func. For example I consider it an anti-pattern to ever set externs on the result of a generator inside the generator itself.

There is currently a bug in the code in this branch that if realize is called multiple times with different externs, each call will use the first set of externs passed in and not notice that they've changed. Comments on whether this is worth fixing are welcome. (The fix likely involves fingerprinting the externs structure and storing multiple compiled versions of the code in the Func structure. We could just move to a cross Func compilation cache as well....)

Or we could head down the path of having an execution context structure for Halide and start pushing upward in terms of framework level. (E.g. the compilation cache idea above would hang off the execution context and starts to provide application level support for using JIT.)

**Comment by dsharletg** [comment]
> After the pending Pipeline changes, the state would be carried on Pipeline, not Func. I'm not sure this materially changes anything, but Pipeline may be enough of an API evolution to consider killing all the handlers and moving to just setting externs on Pipeline.

Doesn't my suggestion regarding setting the externs vs. passing them to realize apply equally to Func and Pipeline?

> hey're both useful, though I'd probably steer people toward bundling up all the deps into a compilation step as opposed to trying to provide them when building the Func. For example I consider it an anti-pattern to ever set externs on the result of a generator inside the generator itself.

That simply sounds like an outright error, and that all the compile_to_* static compilation methods should just assert that there aren't any extern function pointers set. That solves the problem for both generators and otherwise.

> There is currently a bug in the code in this branch that if realize is called multiple times with different externs, each call will use the first set of externs passed in and not notice that they've changed. Comments on whether this is worth fixing are welcome. (The fix likely involves fingerprinting the externs structure and storing multiple compiled versions of the code in the Func structure. We could just move to a cross Func compilation cache as well....)

This is exactly what I was afraid of with the 'pass a bundle' API, and it seems trivial to fix if you just use a stateful API: simply invalidate the cached stuff whenever `add_extern_definition` is called.


I get that stateful things are not nice and I dislike them as well, but I think it makes sense here.

**Comment by zvookin** [comment]
As the code stands, the JITExtern uses pointers to refer to it's target to avoid copying entire trees. The C function pointers have to be valid for as long as the compiled code might be called since they point to code, but the pointers to Funcs only have to be valid during the call to realize, etc. If we set externs as persistent state on a Func, there is more opportunity for the Func pointer in a JITExtern to go stale.

Otherwise I don't really have an objection to changing this to setters on Func. I think there is an overall better design to be had. Perhaps by introducing a JITContext object that holds state like this, but I'm not really up for designing that now.

Andrew, any opinions?

**Comment by abadams** [comment]
Holding Funcs by value doesn't seem like it would be a problem - they're just reference counted pointers, so I don't think it would copy entire trees of anything. In fact that seems superior to avoid any lifetime problems.

Unlike set_error_handler, this would be state that needs to be set before compilation (or it would have to invalidate cached compiled stuff). So it's not quite identical to those functions. I'd be tempted to make it an extra argument to compile_jit *only* so that it's clear from the API that if you want to change it you need to recompile. That would also cut down on argument clutter for realize.

**Comment by zvookin** [comment]
I reworked this on top of the new Pipeline support and it is cleaner now. JITExterns are now set on a pipeline per Dillon's suggestion. I like this a lot better than setting them on a Func and the API does not support setting them on a Func. One must wrap the Func in a pipeline to be able to JIT with explicitly provided externs. I did it this way because all Funcs in a Pipeline have to be compiled with the same externs anyway and the only way that concept exists is if the externs are a property of a Pipeline, not the Funcs.

There are two TODO's which are invitations for comment. One is in Pipeline::set_jit_externs and is asking whether it is worth checking if a state setting operation is doing a nop set or not. The set invalidates the cached compile so it might save a fair bit of performance, but it seems pretty rare this will happen.

The other one is in JITModule.h around the ScalarOrBufferT and ExternSignature types. It seems like we have a lot of Types that are sorta kinda similar in this area. Not sure there's any easy to exploit commonality without some refactoring however.

**Comment by dsharletg** [comment]
How about having `set_jit_extern` and `clear_jit_externs` (and maybe even
`remove_jit_extern`, taking only a name?) instead of `set_jit_externs`?
This has a number of advantages:

- Whether or not to invalidate the cache becomes very simple (not a map
comparison).
- It's more painful/less efficient to construct a map just to call this
function, especially if the JIT externs aren't necessarily all part of the
same code. If you need to just add a jit extern to a pipeline, you need to
copy the existing jit externs, add the new JIT extern, and then set them
again. On the other hand, it's pretty easy to bust up a map of JIT externs
and call set_jit_extern one by one.

**Comment by zvookin** [comment]
Did you look at how this is actually used in the tests?

**Comment by zvookin** [comment]
"- Whether or not to invalidate the cache becomes very simple (not a map
comparison)."

If you have N externs and you are setting M new ones what is the complexity of each design?

With your proposal, I don't see how there is ever a case where the cache survives if one clears the map and rebuilds it from scratch, which will generally be required in any real use.

It isn't a question of difficulty writing the code or efficiency, it is a question of how we want the API to behave. In practice I don't think the reuse case ever occurs because a real app that is exploiting this will need to cache the Pipeline object at a higher level and it will basically be immutable. If things change, it will be rebuilding a new Pipeline anyway. Invalidating the cache is great for correctness and uses like the tests, but I don't see the current API solving general code caching in a real application. As such I'm going to remove the comment and leave the code as is.

**Comment by zvookin** [comment]
Also, this design is consistent with a Pipeline getting its entire list of Funcs at construction time. We do not allow incrementally adding and removing Funcs from a Pipeline.

Generally this design seems much better to me. If I felt distributing the construction of the externs list was a good idea, I'd probably look to carrying them on the Funcs that go into the Pipeline and merging the list together. But that has all sorts of problems. E.g. what happens when two components in this distributed code set the same name to two different things?

**Comment by dsharletg** [comment]
Here's my main motivation behind this feedback, which I'd like to reiterate
from the previous round of review (easy cache invalidation, API usage
simplicity, etc. are only in addition to this): This feels very similar to
set_error_handler, set_do_par_for, etc., eventually we might even be able
to get rid of all the crazy callback special case stuff for the JIT in the
runtime and just rely on the weak linkage mechanism, by simply doing

    set_jit_extern("halide_error", my_error_handler);
    set_jit_extern("halide_do_par_for", my_do_par_for);

If you agree that these are similar, then I think that set_jit_extern is
more conceptually consistent with the existing APIs (set_error_handler,
set_do_par_for, ...), which it would be co-existing with.

I don't use the JIT enough to comment from experience regarding application
design with the JIT. My hunch is that applications will want to call
set_error_handler, set_do_par_for, etc. in a centralized core part of the
application, while "plugins" will want to be calling set_jit_extern from
anywhere. I know that currently, they aren't the same thing, so this is
barely relevant. But, I re-iterate that they *seem* like the same thing,
and hopefully will be relevant eventually (to eliminate some of the extra
logic in the runtime for the JIT).

Anyways, I don't feel that strongly about this either way, and we can
always add the extra APIs later if the refactoring I've described comes
about and makes sense.

**Comment by zvookin** [comment]
"If you agree that these are similar..."

The main goal is to pass a library of functionality into a compilation to satisfy the dependencies of introduced by HalideExtern and define_extern. NamedJITExterns is a library. Packaging it into a data structure allows encapsulating what is in that library, e.g. by having a routine that returns a single map for many Pipeline compilations or other mechanisms. It also happens to be syntactically clean to create a literal one on the fly and is arguably more readable if you are defining more than one extern.

Hander setting works differently as changing them does not require recompilation. So getting rid of that mechanism requires some thought on whether that functionality is useful.

File: src/Pipeline.h [hunk 2]
```diff
@@ -90,6 +90,10 @@ struct Outputs {
     }
 };
 
+struct JITExtern;
+
+typedef std::map<std::string, JITExtern> NamedJITExterns;
```
Comment (abadams) [inline src/Pipeline.h]: I'd mildly prefer not having this typedef, and making it more explicitly a map in the Pipeline API.

File: src/Pipeline.h [hunk 3]
```diff
@@ -303,6 +310,16 @@ class Pipeline {
      */
     EXPORT void set_custom_print(void (*handler)(void *, const char *));
 
+    /** Install a set of external C functions or Funcs to satisfy
+     * dependecies introduced by HalideExtern and define_extern
```
Comment (abadams) [inline src/Pipeline.h]: typo: dependencies

**Comment by abadams** [comment]
I had two trivial comments, but the design LGTM as is. In a complex JIT app I suspect you'd be likely to build up one big map of all your extern junk and pass it to every pipeline you make. Having to iterate over that map and insert the thing one element at a time seems tedious.

**Pull request merged**